[JENKINS-55927] Hook event should not trigger Branch Indexing

Dohbedoh opened the pull request:
[JENKINS-55927](https://issues.jenkins.io/browse/JENKINS-55927)

Noticed this issue while troubleshooting an environment where Bitbucket event were triggering branch indexing. Haven't yet identified exactly what kind of push events has empty `changes`. Though per my understanding of the design of the Branch API / SCM API and Pipeline Multibranch, SCM events should not trigger branch indexing ? And this is rather unexpected. Unless I am missing something ?

cc @jenkinsci/bitbucket-branch-source-plugin-developers 

### Your checklist for this pull request

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Did you provide a test-case? That demonstrates feature works or fixes the issue.

nfalco79 commented:
I think, but I need to check, Approved, Approved Removed, Declined event could have changes. For example declined will trigger the job deletion in multibranch pipeline

Dohbedoh commented:
@nfalco79 Actually according to the code and the HookProcessor that do this, it should really be a push event (could also be a mirror synchronized) but not pull requests... Unless declining a pull requests somehow generate a push event on the base branch..

> For example declined will trigger the job deletion in multibranch pipeline

Right, so per my understanding of the [SCM API](https://github.com/jenkinsci/scm-api-plugin/blob/master/docs/implementation.adoc), a decline should just be a `REMOVED` event and mark the branch job as dead. Branch Indexing will be responsible later on for the cleanup of orphaned items.

* https://github.com/jenkinsci/scm-api-plugin/blob/698.v8e3b_c788f0a_6/docs/implementation.adoc?plain=1#L153-L157
* https://github.com/jenkinsci/scm-api-plugin/blob/698.v8e3b_c788f0a_6/docs/implementation.adoc?plain=1#L1945-L1950

Here the behavior of triggering a full scan on a SCMHead event (single branch) can be detrimental. A full scan can be long and expensive (resources and BranchIndexing thread pool on the server but also API requests against Bitbucket).

That is why I think this might be a mistake.

nfalco79 commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/PushHookProcessor.java:
```diff
@@ -71,13 +70,6 @@ public void process(HookEventType hookEvent, String payload, BitbucketType insta
                 push = BitbucketCloudWebhookPayload.pushEventFromPayload(payload);
             }
             if (push != null) {
-                String owner = push.getRepository().getOwnerName();
```
As described in JIRA issue I think these changes will cause to the user that have setup "Discover pull requests" strategies different to "The current pull request revision" to not re-build anymore the PRs target the same merged branch of the event.
If I understand correct the configuration in the JIRA issue the reported would trigger manual reindex to update all PRs. Than this should be archived using a build strategies to ignore some specific index events (maybe already implemented in other plugins)

nfalco79 commented:
> Here the behavior of triggering a full scan on a SCMHead event (single branch) can be detrimental. A full scan can be long and expensive (resources and BranchIndexing thread pool on the server but also API requests against Bitbucket).

I agree but if you do not perform a BranchIndexing you do not know what are branches involved in changes.

Dohbedoh commented:
Looking at the history of that change, it was here from the start, before we implemented the SCM API Event mechanism and preserved when we added it https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/37/files#diff-a5efe0e74b0f75735db75b177d0b8da3426c13be4a855727e8eeb8193f74ff33. Though we were relying on a custom made plugin at that time that does not exist anymore: https://github.com/topicusfinan/bitbucket-webhooks-plugin/

I think we really need to figure out what scenario causes an empty `changes`. I am quite convinced triggering indexing instead of using a particular Event is wrong.. If we wanted to have indexing triggered, I would think we'd create a SCMSourceEvent instead. Maybe someone closer to the design of SCM API / Branch API could confirm cc @rsandell @amuniz 

Per the JIRA issue, this could be a remote merge, that is we have an Open PR, but instead of merging it from Bitbucket, we merge it by hand. Hence we end up with a PR without changes. If that is the case, this really shouldn't trigger an indexing. I would be surprised if someone was relying on such a behavior.

nfalco79 commented:
I would take time to test some supported webhook event (I'm collecting them) like:
new tag, that must automatically trigger the tag discover branch
PR merged trigger removal
New PR
Approved
Disapproved

The behaviour to trigger a new build for all the PRs with Merge strategy is by what who setup this expect to ensure the PR is good for merge and avoid trigger manul build. Who do not want -> build strategy to inhibit builds (but not indexing)

amuniz commented:
It's a really long time I worked on this plugin, so I don't really know. But looking at the [change introducing the empty check](https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/37/files#diff-a5efe0e74b0f75735db75b177d0b8da3426c13be4a855727e8eeb8193f74ff33) I think they were interpreted as an empty changeset, so it corresponds to a PR creation/deletion which actually requires a re-index to create/remove the job in Jenkins. Whether that's still needed or not, I don't know, but I think if all the known cases are manually tested and they work, then we are ok.

Dohbedoh commented:
But per the design of the API (and like all implementation do) those events do not run branch indexing... A CREATE SCMevent would create the PR. A DELETED SCMEvent would mark it as stale... And those event with empty `changes` are actually push events, so it actually impacts branches mainly.

Just need to figure out what a push event with empty changes originate from. I think I have a reproducer with Bitbucket Server with Native webhook and also with the add on. Will update this in the JIRA shortly.

Dohbedoh commented:
Found 2 scenario where you get a push event with empty {{changes}}, in Bitbucket Server with the [Native webhook](https://confluence.atlassian.com/bitbucketserver/keeping-forks-synchronized-776639961.html) as well as with the Bitbucket add-on. I found 2 scenarios where the empty changes may happen:

**Rebase a target branch to a source branch while a PR is open**

```
git checkout main
git checkout -B source-1
echo "test" >> source-1.md
git add . && git commit -m "source-1: $(date)" && git push --set-upstream origin source-1
git push --set-upstream origin source-1
# Open a PR for from the source branch `source-1` to the target branch `main`
git checkout main
git rebase -i source-1
git push -f
```

This will generate 2 events. The first one with empty `changes` and the next one with the correct changes made to the target branch. (The PR is automatically merged because there is no diff and that's a default behavior of Bitbucket server).. So the first event is useless. It does not give any clue about what just happened. 

**Auto-Sync of Fork when a branch cannot be synchronized**

When the Multibranch Pipeline is a fork and [Automatic Fork Syncing](https://confluence.atlassian.com/bitbucketserver/keeping-forks-synchronized-776639961.html) is enabled. If a branch on a fork cannot be synchronized, the attempt from Bitbucket to synchronize it generates a push event with empty `changes`. This seems also useless. The event does not contain information that allow us to know what just happened, and the refs is actually unchanged...

****

Both the add-on and the native webhook results behave the same. Per my understanding the add-on relies on the native mechanism but add some extra configuration.. 

So far, I don't see a valid scenario where a push event with empty changes should be considered. Or why it would trigger branch indexing. If we are skeptical about whether anybody expect this behavior, I propose to add a system property to allow use to re-enable this.

nfalco79 commented:
>The behaviour to trigger a new build for all the PRs with Merge strategy (build PR commit with merge into master) is by what who setup this expect to ensure the PR is good for merge and avoid trigger manual build. 

Issue #316

Dohbedoh commented:
Actually this behavior is not what happens with Bitbucket Branch Source at the moment. Those events that trigger branch indexing are happening on very specific scenario where you would not expect it (see my previous comment). 
What people are asking in https://github.com/jenkinsci/bitbucket-branch-source-plugin/issues/316 are IMO a feature for SCM API / Branch API. By default the API does not go and observe source branches when a target one is updated.

Dohbedoh commented:
I still don't have an answer from Atlassian at https://community.atlassian.com/t5/Bitbucket-questions/Webhook-repo-refs-changed-events-with-empty-changes-array/qaq-p/2868488#M109342. But note that the plugin maintained by Atlassian only move forward if the event contains changes:

* https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/blob/7da79af5f2f137ed38b1f31828b45f935a9d86f4/src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java#L59-L61
* https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/blob/7da79af5f2f137ed38b1f31828b45f935a9d86f4/src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java#L101C44-L106

Maybe that can help convinced maintainers [JENKINS-55927](https://issues.jenkins.io/browse/JENKINS-55927) is truly a bug and this fix is valid.

Dohbedoh commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/HookProcessor.java:
```diff
@@ -86,6 +85,7 @@ public void process(HookEventType type, String payload, BitbucketType instanceTy
     /**
      * To be called by implementations once the owner and the repository have been extracted from the payload.
      *
+     * @deprecated SCM Event should not trigger Branch Indexing
```
The impact really depend on a the environment. User that have large repositories with hundreds of branches / PRs cannot afford frequent branch indexing.
In case of #539, branch indexing is not a good solution IMO. A pull request decline event should not trigger branch indexing, but rather update the source and target based on discovery criterias.

nfalco79 replied:
> but rather update the source and target based on discovery criteria through SCMHeadEvents.

This is hard because you should apply any registered trait (also external to this plugin) and simulate if any existing job is related to the target branch and in case delete or create.

nfalco79 commented on src/test/resources/com/cloudbees/jenkins/plugins/bitbucket/hooks/server/emptyPayload.json:
Is this a real payload webhook?

Dohbedoh replied:
Nope, taken from https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/master/src/test/resources/com/cloudbees/jenkins/plugins/bitbucket/client/events/BitbucketCloudPushEventTest/emptyPayload.json. I don't have a non native server instance to test and get a payload.
Maybe I can use https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/master/src/test/resources/com/cloudbees/jenkins/plugins/bitbucket/server/events/BitbucketServerPushEventTest/updatePayload.json and change to an empty array of changes.

nfalco79 replied:
I would real example (masked) instead to provide payload that support what would do.

nfalco79 commented:
Further details about server empty changes:
https://docs.atlassian.com/bitbucket-server/javadoc/5.7.0/api/reference/com/atlassian/bitbucket/event/repository/RepositoryRefsChangedEvent.html#getRepository()

Note: While the returned collection will never be null, it may be empty. For example, a [remotely merged pull request](https://docs.atlassian.com/bitbucket-server/javadoc/5.7.0/api/reference/com/atlassian/bitbucket/event/pull/PullRequestMergedEvent.html#isMergedRemotely()) will have an empty set of changes.

Dohbedoh commented:
@nfalco79 Yes, as mentioned above and in the JIRA. In that case, you receives 2 events. One without changes and the second one with the commits from the remote merge. The second can just be discarded.

jtnord commented:
> Haven't yet identified exactly what kind of push events has empty `changes`

A mirror synchronized event with too mnay refs to list will set `refLimitExceeded = true` and have an empty changes.

http://confluence.atlassian.com/bitbucketserver/event-payload-938025882.html#Eventpayload-Mirrorsynchronized

Dohbedoh commented:
Thnaks @jtnord. I have update the PR and added tests for this scenario.

jtnord commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/NativeServerPushHookProcessor.java:
```diff
@@ -44,6 +44,9 @@
 @RestrictedSince("933.3.0")
 public class NativeServerPushHookProcessor extends HookProcessor {
 
+    private static final boolean SCAN_ON_PUSH_WITH_EMPTY_CHANGES = Boolean.getBoolean(
+        NativeServerPushHookProcessor.class.getName() + ".scanOnPushWithEmptyChanges");
```
NIT: Use a [SystemProperty](https://javadoc.jenkins.io/jenkins/util/SystemProperties.html)?  (not everyone can set System properties).
Should this default to false (ie invert the setting so you disable full scans for empty changes rather than the other way around?)

Dohbedoh replied:
Done.

jtnord commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/events/NativeServerMirrorRepoSynchronizedEvent.java:
```diff
@@ -31,8 +31,11 @@ public class NativeServerMirrorRepoSynchronizedEvent {
+
     private List<NativeServerChange> changes;
 
+    private Boolean refLimitExceeded;
```
why Boolean not a boolean  is this not required in the payload? (I see it in the datacenter 7.21 and 9.5 docs) (the whole event seems to be specific to BB Datacenter and not BB Cloud)?

jtnord reviewed (commented):
Seems reasonable, however I would wonder if the boolean should be inverted so that you opt into the new behaviour rather than out of it.

At the end of the day triggering a scan should be fine, and if that is an issue, then scanning should be improved to rate limit or coalesce full repository scans.

nfalco79 commented on src/test/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/NativeServerPushHookProcessorTest.java:
```diff
@@ -0,0 +1,104 @@
+
+        ArgumentCaptor<ServerPushEvent> eventCaptor = ArgumentCaptor.forClass(ServerPushEvent.class);
+        verify(sut).notifyEvent(eventCaptor.capture(), anyInt());
+        ServerPushEvent event = eventCaptor.getValue();
```
you are not testing the core method of ServerPushEvent#heads or #getPullRequests

nfalco79 commented on src/test/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/PushHookProcessorTest.java:
```diff
@@ -32,36 +32,39 @@
-                PushHookProcessorTest.this.scmEvent = event;
-            }
-        };
+        sut = spy(new PushHookProcessor());
```
revert changes to avoid spy

Dohbedoh commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/PushHookProcessor.java:
```diff
@@ -52,15 +55,20 @@ public void process(HookEventType hookEvent, String payload, BitbucketType insta
-                    LOGGER.log(Level.INFO, "Received hook from Bitbucket. Processing push event on {0}/{1}",
+                    final String owner = push.getRepository().getOwnerName();
+                    final String repository = push.getRepository().getRepositoryName();
+                    if (SCAN_ON_PUSH_WITH_EMPTY_CHANGES) {
```
or you rather mean `BitbucketApiUtils.isCloud(push.getLinks().get("self")) || SCAN_ON_PUSH_WITH_EMPTY_CHANGES` ? Do reindexing only is this is Bitbucket Cloud OR if `SCAN_ON_PUSH_WITH_EMPTY_CHANGES` is `true`.

nfalco79 replied:
My fault I mean if bitbucket cloud than do as usual, in case of bitbucket server than take care of property

Dohbedoh commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/HookProcessor.java:
```diff
@@ -88,6 +87,7 @@ public void process(HookEventType type, String payload, BitbucketType instanceTy
     /**
      * To be called by implementations once the owner and the repository have been extracted from the payload.
      *
+     * @deprecated Branch Indexing should not be triggered directly. But through {@link jenkins.scm.api.SCMSourceEvent}.
```
I wanted to limit the changes to just address the problem for now. And speed up the merge of this PR. Instead of deprecating the method, I added a TODO comment over the usage of the deprecated method `scmOwner.onSCMSourceUpdated`.

Dohbedoh commented on src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/HookProcessor.java:
```diff
@@ -100,8 +100,11 @@ protected void scmSourceReIndex(final String owner, final String repository) {
                             && scmSource.getRepoOwner().equalsIgnoreCase(owner)
-                            && scmSource.getRepository().equals(repository)) {
+                            && scmSource.getRepository().equals(repository)
+                            && (mirrorId == null || mirrorId.equalsIgnoreCase(scmSource.getMirrorId()))) {
```
@nfalco79 I noticed that this condition was missing. This could cause indexing on a primary (non mirror) repository multibranch when a mirror is syncrhonized. Which per my understand is unexpected.

Dohbedoh commented on docs/USER_GUIDE.adoc:
```diff
@@ -257,9 +257,26 @@ For the same reason existing projects must be updated with a "Scan Organization
+
+This behavior can be disabled with the following system properties:
+
+* `com.cloudbees.jenkins.plugins.bitbucket.hooks.PushHookProcessor.scanOnEmptyChanges=false` for Cloud and Server (non-native)
```
Got it. Added the property definition in the `HooKProcessor` for now.

nfalco79 commented:
@Dohbedoh could you provide a single commit with a description of changes ?
